# **Numerical Computing with Numpy**
NumPy (Numerical Python) is a module which was created allow efficient numerical calculations on multi-dimensional arrays of numbers from within Python. It is derived from the merger of two earlier modules named Numeric and Numarray. The actual work is done by calls to routines written in the Fortran and C languages.

- NumPy defines a new data type called the `ndarray` (n-dimensional array).
- The elements of ndarrays can be integers, floats, and complex numbers of various sizes. 
- When the NumPy package is loaded, ndarrays become as much a part of the Python language as standard Python data types such as lists and dictionaries.
- A NumPy array can only hold one single data type at a time.

##### **To install Numpy, run this command**

In [ ]:
# pip install numpy

##### **To load numpy, import the installed numpy module**

In [2]:
import numpy as np

### **1. Creating arrays**

In [10]:
a = np.array([[1, 2, 3], [4, 5, 6]])
a

array([[1, 2, 3],
       [4, 5, 6]])

In [11]:
a.dtype

dtype('int32')

In [12]:
b = np.array([[1., 2., 3.], [4., 5., 6.]])
b

array([[1., 2., 3.],
       [4., 5., 6.]])

In [15]:
b.dtype

dtype('float64')

In [17]:
b.shape

(2, 3)

### **2. Creating one-dimensional sequences**
The `arange` function creates a one-dimensional array consisting of a sequence of numbers.

Syntax:
```bash
array = np.arange(start, stop, step)
```

In [18]:
b = np.arange(0,11)
b

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

In [19]:
c = np.arange(0.,2.,0.4)
c

array([0. , 0.4, 0.8, 1.2, 1.6])

**Example**: Inaccuracy in floating point arithmetic with arange function

In [21]:
d = np.arange(0.,2.1,0.3)
d

array([0. , 0.3, 0.6, 0.9, 1.2, 1.5, 1.8, 2.1])

**Alternative Method**: `linspace` function

Syntax: 
```bash
array = np.linspace(start, stop, number)
```

Trả về một mảng gồm số phần tử được phân bố đều với các giá trị bắt đầu và kết thúc được chỉ định.

In [27]:
d = np.linspace(0.,2.1,3)
d

array([0.  , 1.05, 2.1 ])

In [28]:
e = np.linspace(0.,3.5,8)
e

array([0. , 0.5, 1. , 1.5, 2. , 2.5, 3. , 3.5])

### **3.  The `zeros` and `ones` functions**

In [29]:
a = np.zeros((2,4))
a

array([[0., 0., 0., 0.],
       [0., 0., 0., 0.]])

In [31]:
b = np.ones((3,3,3))
b

array([[[1., 1., 1.],
        [1., 1., 1.],
        [1., 1., 1.]],

       [[1., 1., 1.],
        [1., 1., 1.],
        [1., 1., 1.]],

       [[1., 1., 1.],
        [1., 1., 1.],
        [1., 1., 1.]]])

In [35]:
c = np.ones((2,3,4))
c

array([[[1., 1., 1., 1.],
        [1., 1., 1., 1.],
        [1., 1., 1., 1.]],

       [[1., 1., 1., 1.],
        [1., 1., 1., 1.],
        [1., 1., 1., 1.]]])

In [36]:
d = np.zeros((2, 3, 3, 4))
d

array([[[[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]],

        [[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]],

        [[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]]],


       [[[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]],

        [[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]],

        [[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]]]])

### **4. Array data types**


##### ́**4.1. Integer**
- Signed Integer: The number trailing the type indicates the bit-width. The more bits you use, the larger the numbers you can store, but the more RAM you consume.
    - int8: Ranges from -128 to 127.
    - int16: Ranges from -32,768 to 32,767.
    - int32: Ranges from $-2.14 \times 10^9$ to $2.14 \times 10^9$.
    - int64: Ranges from $-9 \times 10^{18}$ to $9 \times 10^{18}$.
- Unsigned Integer: By dropping the negative numbers, unsigned integers shift their entire mathematical range into the positive zone.
    - uint8: Ranges from 0 to 255.
    - uint16: Ranges from 0 to 65,535.
    - uint32: Ranges from 0 to $4.29 \times 10^9$.
    - uint64: Ranges from 0 to $18.4 \times 10^18$.

In [9]:
a = np.array([1, 2], dtype = np.int8)
a

array([1, 2], dtype=int8)

In [19]:
b = np.array([-11, 2], dtype = np.uint8)
b

C:\Users\phaml\AppData\Local\Temp\ipykernel_20904\1973994269.py:1: DeprecationWarning: NumPy will stop allowing conversion of out-of-bound Python integers to integer arrays.  The conversion of -11 to uint8 will fail in the future.
For the old behavior, usually:
    np.array(value).astype(dtype)
will give the desired result (the cast overflows).
  b = np.array([-11, 2], dtype = np.uint8)


array([245,   2], dtype=uint8)

##### **4.2. Float**
- float16: $\pm 6.55 \times 10^4$
- float32: $\pm 3.40 \times 10^{38}$
- float64: $\pm 1.80 \times 10^{308}$

In [21]:
a = np.array([1, 2], dtype = np.float64)
a

array([1., 2.])

##### **4.3. Complex (số phức)**
- complex64: $\pm 3.40 \times 10^{38}$
- complex128: $\pm 1.79 \times 10^{308}$
- complex256: $\pm 1.18 \times 10^{4932}$

In [22]:
a = np.array([2,3,4], dtype = np.complex64)
a

array([2.+0.j, 3.+0.j, 4.+0.j], dtype=complex64)

In [24]:
a = np.array([2,3,4], dtype = np.complex128)
a

array([2.+0.j, 3.+0.j, 4.+0.j])

In [23]:
b = np.ones((2,3), dtype = np.complex64)
b

array([[1.+0.j, 1.+0.j, 1.+0.j],
       [1.+0.j, 1.+0.j, 1.+0.j]], dtype=complex64)

##### **4.4. `astype` method**

In [25]:
a = np.arange(0,5)
a

array([0, 1, 2, 3, 4])

In [26]:
a.dtype

dtype('int32')

In [27]:
b = a.astype('float32')
b

array([0., 1., 2., 3., 4.], dtype=float32)

In [30]:
c = a.astype('float64')
c

array([0., 1., 2., 3., 4.])

### **5. Array indexing and looping**

##### **5.1. Basic indexing**

In [31]:
a = np.array([[2.1, 3.5, 7.0], [10.4, 12.9, 5.1]])
a

array([[ 2.1,  3.5,  7. ],
       [10.4, 12.9,  5.1]])

In [32]:
a[0,0]

2.1

In [33]:
a[1,2]

5.1

##### **5.2. Slicing**

**Example 1**:

In [34]:
a = np.array([[2.1, 3.5, 7.0], [10.4, 12.9, 5.1]])
a

array([[ 2.1,  3.5,  7. ],
       [10.4, 12.9,  5.1]])

In [35]:
b = a[:, 1:2]
b

array([[ 3.5],
       [12.9]])

- The index “:” returns all elements along the corresponding row. 
- The index "1:2" returns the second element of each row.

In [36]:
c = a[:]
c

array([[ 2.1,  3.5,  7. ],
       [10.4, 12.9,  5.1]])

**Example 2**:

In [37]:
a = np.zeros((2,2))
a

array([[0., 0.],
       [0., 0.]])

In [38]:
b = np.array([1., 2.])
b

array([1., 2.])

In [40]:
a[0, :] = b
a

array([[1., 2.],
       [0., 0.]])

- "0": Targets Row 0 
- ":": Targets all columns in that row
- "= b": Copies the values from b directly into those targeted slots

**Example 3**:

In [41]:
a = np.arange(0,10)
a

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [42]:
a[0:10:2]

array([0, 2, 4, 6, 8])

In [44]:
a[-1:0:-1]

array([9, 8, 7, 6, 5, 4, 3, 2, 1])

In [47]:
a[-1::-1]

array([9, 8, 7, 6, 5, 4, 3, 2, 1, 0])

##### **5.3. Viewing and copying**

**Viewing**

In [50]:
a = np.array([[1, 2], [3, 4]])
a

array([[1, 2],
       [3, 4]])

In [51]:
b = a[0:1,1:2]
b

array([[2]])

In [52]:
a[0,1] = 32
b

array([[32]])

**Copying**: To make a new array out of a slice of the original array, use of the “+” prefix.

In [53]:
a = np.array([[1, 2], [3, 4]])
a

array([[1, 2],
       [3, 4]])

In [54]:
b = +a[0:1,1:2]
b

array([[2]])

In [55]:
a[0,1] = 32
b

array([[2]])

##### **5.4. Reshaping arrays**
`reshape` method allows the number of dimensions and the size of each dimension to be changed as long as the total number of array elements remains the same.

In [65]:
a = np.array([[1, 2], [3, 4]])
a

array([[1, 2],
       [3, 4]])

In [66]:
a.reshape(4)

array([1, 2, 3, 4])

- The numbers in array `a` are read row-by-row 
- And unpacked into a single horizontal row.

In [67]:
a.reshape(4,1)

array([[1],
       [2],
       [3],
       [4]])

In [68]:
a.reshape(2,2,1)

array([[[1],
        [2]],

       [[3],
        [4]]])

- 2 (First Dimension): Creates 2 separate matrices.
- 2 (Second Dimension): Each of those matrices contains 2 rows.
- 1 (Third Dimension): Each row contains exactly 1 element.

In [69]:
a

array([[1, 2],
       [3, 4]])

In [70]:
b = a.reshape(4)
b

array([1, 2, 3, 4])

In [71]:
a[0,1] = 13
a[1,0] = 45
b

array([ 1, 13, 45,  4])

As with slicing, reshaping doesn’t normally create a new array, just a new view of the original array.

##### **5.5. Array looping**

**1D array**

In [4]:
a = np.arange(0,5)
a

array([0, 1, 2, 3, 4])

In [5]:
for item in a:
    print(item)

0
1
2
3
4


**2D array**

In [6]:
a = np.array([[1, 2], [3, 4]])
a

array([[1, 2],
       [3, 4]])

In [7]:
for row in a:
    print(row)

[1 2]
[3 4]


In [9]:
for row in a:
    for item in row:
        print(item)

1
2
3
4


**`flat` attribute**

In [10]:
a = np.array([[1, 2], [3, 4]])
a

array([[1, 2],
       [3, 4]])

In [12]:
for item in a.flat:
    print(item)

1
2
3
4


**`np.ndenumerate()` method**

In [13]:
a = np.array([[1, 2], [3, 4]])
a

array([[1, 2],
       [3, 4]])

In [15]:
for index, item in np.ndenumerate(a):
    print(f"{index} has value {item}")

(0, 0) has value 1
(0, 1) has value 2
(1, 0) has value 3
(1, 1) has value 4


**`np.nditer()` method**

In [16]:
a = np.array([[1, 2], [3, 4]])
a

array([[1, 2],
       [3, 4]])

In [17]:
for item in np.nditer(a):
    print(item)

1
2
3
4


In [19]:
with np.nditer(a, op_flags=['readwrite']) as iter:
    for item in iter:
        item[...] = item*10

a

array([[10, 20],
       [30, 40]])

In [20]:
a = np.array([[1, 2], [3, 4]])
a

array([[1, 2],
       [3, 4]])

In [21]:
with np.nditer(a, op_flags=['readwrite']) as iter:
    for item in iter:
        item[...] = item*10 if item%2==0 else item

a

array([[ 1, 20],
       [ 3, 40]])

### **6. Array expressions and assignments**

In [24]:
a = np.array([[1, 2], [3, 4]])
a

array([[1, 2],
       [3, 4]])

In [25]:
b = np.array([[5,6],[7,8]])
b

array([[5, 6],
       [7, 8]])

In [23]:
c = a + b
c

array([[ 6,  8],
       [10, 12]])

In [26]:
d = a * b
d

array([[ 5, 12],
       [21, 32]])

In [27]:
e = a*b + c
e

array([[11, 20],
       [31, 44]])

In [29]:
f = c + a*b
f

array([[11, 20],
       [31, 44]])

##### **Broadcasting**

In [30]:
a = np.array([[1, 2], [3, 4]])
a

array([[1, 2],
       [3, 4]])

In [31]:
e = np.array([10, 20, 30])
e

array([10, 20, 30])

In [32]:
a + e

ValueError: operands could not be broadcast together with shapes (2,2) (3,) 

The shape of array a is (2,2),while the shape of array e is (3,), and they therefore cannot be added together.

In [33]:
a.shape

(2, 2)

In [34]:
e.shape

(3,)

In [37]:
b = np.array([[5,10]])
b.shape

(1, 2)

In [38]:
a + b

array([[ 6, 12],
       [ 8, 14]])

In this case array b, which has one row and two columns, is (conceptually) extended along axis 0, repeating values as necessary, to create an array with two rows array([[5,10],[5,10]]) and then added to a. This extension process is called broadcasting.

In [41]:
c = np.array([[5], [10]])
c.shape

(2, 1)

In [42]:
a + c

array([[ 6,  7],
       [13, 14]])

In this case array c is extended along axis 1, to create an array with two rows array([[5,5],[10,10]]), which is then added to a. Notice that the result here is different than that in the previous case, since the broadcast axis is different.

In [43]:
b + c

array([[10, 15],
       [15, 20]])

That is a case of double broadcasting.

Để hai mảng có thể thực hiện broadcasting với nhau, NumPy sẽ so sánh shape của chúng từ phải qua trái (từ chiều cuối cùng lên chiều đầu tiên). Hai chiều được coi là tương thích khi và chỉ khi:
- Chúng bằng nhau.
- Hoặc một trong hai chiều có kích thước bằng 1.

In [48]:
d = np.array([5, 10])
d

array([ 5, 10])

In [49]:
d.shape

(2,)

In [50]:
a + d

array([[ 6, 12],
       [ 8, 14]])

In [51]:
d1 = d.reshape(1,2)
d1

array([[ 5, 10]])

In [52]:
a + d1

array([[ 6, 12],
       [ 8, 14]])

In [53]:
d2 = d.reshape(2,1)
d2

array([[ 5],
       [10]])

In [54]:
a + d2

array([[ 6,  7],
       [13, 14]])